### ISSUE TITLE
Reformer: The Efficient Transformer

### ISSUE THREAD

**Issue by reo11** [reo11/papers100knock#33]
## 概要
- Transformerを効率化したReformerの提案
- Transformerを使ったモデルは大量の計算資源を必要とするが、Locality-Sensitive HashingとReversible Transformerを使って解決する
- Complexity（計算コスト）をO(L^2)からO(LlogL)に改善

## 投稿年/会議
2020年/ICLR2020
- https://openreview.net/forum?id=rkgNKkHtvB

## 著者/所属機関
Nikita Kitaev, Lukasz Kaiser, Anselm Levskaya
- Google Research

## 参考リンク
- https://www.acceluniverse.com/blog/developers/2020/02/reformer-the-efficient-transformer.html
- https://blog.hoxo-m.com/entry/2020/04/21/090000
- https://ai-scholar.tech/articles/treatise/reformer-ai-364
- https://marvin.news/8173
- https://github.com/lucidrains/reformer-pytorch

### 既存のTransformer
- Attention weightの計算
![image](https://user-images.githubusercontent.com/38539915/82759038-6db86100-9e25-11ea-988a-d121a7789818.png)
- 大きいモデルだと載せるだけで10GB以上のメモリが必要
- しかし、softmax(QK^T)を計算する時に重要なのは大きい値だけなので、queryに近いkeyに対してだけ計算を行なえばよい
  - どうやってその近いkeyを取ってくるのか？→LSH

### Locality Sensitive Hashing（LSH）
- 高次元データを確率的に次元圧縮する手法[[slide](https://www.slideshare.net/yaruki_nil/lsh)]
- LSHの例
- 球面に配置された2つの点は隣接していない限り、3面における回転で全て一致することはない
![image](https://user-images.githubusercontent.com/38539915/82759012-29c55c00-9e25-11ea-87b1-370cb83a73ee.png)

- Q=Kとして、それぞれハッシュし、近傍bucketでchunkする
- LSHを用いたAttention
![image](https://user-images.githubusercontent.com/38539915/82759431-5cbd1f00-9e28-11ea-97da-0c22ee1e5071.png)
- Attentionの処理を同じchunkとそのひとつ前のchunkに対してのみ行うことで計算量を大幅に減らすことができる
![image](https://user-images.githubusercontent.com/38539915/82760559-21721e80-9e2f-11ea-9389-341480150db8.png)

- O(n^2)→O(nlogn)
- これで計算量は改善できたが、学習時に逆伝播するためのActivationを保持しておかねばならず、長い系列だとメモリ消費が激しい問題がある→RevNet

### Reversible Transformer
- Reversible Transformerでは、順伝播時にActivationを保持せず、逆伝播時に再計算する方法を取った（#34 のReversible Residual Networkを採用）
![image](https://user-images.githubusercontent.com/38539915/82760395-3e5a2200-9e2e-11ea-874b-337aa264fcdb.png)
- 通常のNNは(a)のようになっている
- RevNetでは、(b)のように順伝播し、(c)のように逆伝播できる模様
- また、chunkごとに分けて計算できるので、レイヤが増えても大丈夫
- これによってActivationを保持しておく必要がなく、メモリを削減できる
![image](https://user-images.githubusercontent.com/38539915/82760706-5f237700-9e30-11ea-9a9c-b16afa05092b.png)

### 実験
- 入力サイズが大きい二つのタスクに対して
  - enwik8（テキストタスク）：入力トークン64k
  - imagenet64（画像生成タスク）：入力トークン12k
- QKをshareする場合としない場合だと、enwikではshareした方が良い結果
- また、通常のTransformerとActivationをReversibleにしたTransformerでは結果がほぼ同じになっているのがわかる
![image](https://user-images.githubusercontent.com/38539915/82761058-a90d5c80-9e32-11ea-8060-7f948c05655f.png)

- newstest2014 WMT En-Deタスクでもスコアが悪化せず、計算量を削減
![image](https://user-images.githubusercontent.com/38539915/82761121-318bfd00-9e33-11ea-8b19-c173b7f8754f.png)

- LSH Attentionによってスコアがどのように変化するか
![image](https://user-images.githubusercontent.com/38539915/82761144-5d0ee780-9e33-11ea-9879-9a71eb87b06c.png)
- ハッシュする回数を8以上にするとほぼ同様のスコア
 - LSHは入るバケットが確率で決定する為、類似した要素が違うバケットに入る可能性があるため複数回行なうことでfull attentionに近づく

- 左図：レイヤ数を増やすことでスコアが改善する
- 右図：入力シーケンスを伸ばした時に各ステップでかかる時間を表す（対数グラフ）
![image](https://user-images.githubusercontent.com/38539915/82761178-af500880-9e33-11ea-9e1f-25b4185526b8.png)
- 短いシーケンスに対しては微妙だが、系列が長くなるほどTransformerより学習が早い
